src/hotspot/share/services/memTracker.hpp

```diff
@@ -267,6 +267,7 @@ class MemTracker : AllStatic {
     assert_post_init();
     if (tracking_level() < NMT_summary) return;
     if (addr != NULL) {
+      ThreadCritical tc;
```

Comment:
With/without ThreadCritical lock, the assertion in (new|delete)_thread_stack is unreliable, because tracking level is undergraded without ThreadCritical lock, therefore, there is no synchronize-with relationship. So to check tracking level inside (new|delete)_thread_stack, it can only rely on 

Early return at line #268 avoids very expensive ThreadCritcal lock, given NMT is off by default and also tracking level can only be downgraded monotonically.

Anyway, I believe current shutdown logic for virtual memory tracking is racy, filed JDK-8277788.

Reply:
Also would like one less use of ThreadCritical lock, since it's used for malloc if needed for resource allocation.